File: src/BinaryTreeTest.java
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import org.junit.jupiter.api.Test;

class BinaryTreeTest {

    @Test
    public void testEmptyTree() {
        BinaryTree tree = new BinaryTree();
        assertNull(tree.findLCA(0, 0));
    }

    @Test
    public void testTreeWithOneNode() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        assertEquals(1, tree.findLCA(1,1).data);
    }

    @Test
    public void testTree() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        assertEquals(1, tree.findLCA(2,3).data);
    }

    @Test
    public void testLeftTree() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.left.left = new Node(3);
        assertEquals(2, tree.findLCA(2,3).data);
    }

    @Test
    public void testRightTree() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.right = new Node(2);
        tree.root.right.right = new Node(3);
        assertEquals(2, tree.findLCA(2,3).data);
    }

    @Test
    public void testSubTreeLeft() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);
        tree.root.right.left = new Node(6);
        tree.root.right.right = new Node(7);
        assertEquals(2, tree.findLCA(4,5).data);
    }

    @Test
    public void testSubTreeRight() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);
        tree.root.right.left = new Node(6);
        tree.root.right.right = new Node(7);
        assertEquals(3, tree.findLCA(6,7).data);
    }

    @Test
    public void testWithNegatives() {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(-1);
        tree.root.left = new Node(-2);
        tree.root.right = new Node(-3);
        tree.root.left.left = new Node(-4);
        tree.root.left.right = new Node(-5);
        tree.root.right.left = new Node(-6);
        tree.root.right.right = new Node(-7);
        assertEquals(-3, tree.findLCA(-6,-7).data);
    }
}

File: src/BinaryTreeTest.py
import unittest
import BinaryTree


class MyTestCase(unittest.TestCase):
    BinaryTree.root = BinaryTree.Node(1)
    BinaryTree.root.left = BinaryTree.Node(2)
    BinaryTree.root.right = BinaryTree.Node(3)
    BinaryTree.root.left.left = BinaryTree.Node(4)
    BinaryTree.root.left.right = BinaryTree.Node(5)
    BinaryTree.root.right.left = BinaryTree.Node(6)
    BinaryTree.root.right.right = BinaryTree.Node(7)

    def test_node4_5(self):
        self.assertEqual(BinaryTree.findLCA(BinaryTree.root, 4, 5).key, 2)

    def test_node6_2(self):
        self.assertEqual(BinaryTree.findLCA(BinaryTree.root, 4, 6).key, 1)

    def test_node1_5(self):
        self.assertEqual(BinaryTree.findLCA(BinaryTree.root, 3, 4).key, 1)

    def test_node_same(self):
        self.assertEqual(BinaryTree.findLCA(BinaryTree.root, 2, 4).key, 2)

    def test_node2_3(self):
        self.assertEqual(BinaryTree.findLCA(BinaryTree.root, 7, 4).key, 1)

    def test_node_none(self):
        self.assertEqual(BinaryTree.findLCA(None, 4, 5), None)

    def test_node_fake(self):
        self.assertEqual(BinaryTree.findLCA(BinaryTree.root, 8, 8), None)


if __name__ == '__main__':
    unittest.main()
